ocaml-lsp-server/src/document_store.ml

```diff
@@ -1,36 +1,154 @@
 open! Import
 open Fiber.O
 
-type t = (Uri.t, Document.t) Table.t
+type server = Server : 'a Server.t Fdecl.t -> server
 
-let make () = Table.create (module Uri) 50
+type doc =
+  { (* invariant: if [document <> None], then no promotions are active *)
+    document : Document.t option
+  ; (* the number of associated promotions. when this is 0, we may unsubscribe
+       from code actions *)
+    promotions : int
+  }
 
-let put store doc = Table.set store (Document.uri doc) doc
+type t =
+  { db : (Uri.t, doc) Table.t
+  ; server : server
+  ; (* The pool is needed to run subscribe/unsubscribe requests. To prevent
+       deadlocks with synchronous responses to lsp. In the future, these
+       deadlocks should cause runtime errors or will just be impossible *)
```

Comment:
Thanks for the comments! They're useful. 

> In the future, these deadlocks should cause runtime errors or will just be impossible

"In the future" = when some planned changes to Fiber happen?

Reply:
Not to fiber, but either to jsonrpc-fiber or our implementation of LSP on top of it. Basically, it should be forbidden to send requests without responding inside a `later`.